### PR TITLE
Add animations codelab

### PR DESCRIPTION
**Pull request by johnpryan**
This adds the code for the animations codelab, there is still a few steps we would like to add at the end (such as demonstrating how to use OpenContainer), but this is the bulk of it. PTAL, thanks!

cc: @justinmc

### CODE REVIEW

File: animations/step_01/lib/home_screen.dart [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+          children: [
+            Text(
+              '✏️',
+              style: Theme.of(context).textTheme.displayLarge,
+            ),
+            Text(
+              'Flutter Quiz',
+              style: Theme.of(context).textTheme.displayLarge!.copyWith(
+                  fontWeight: FontWeight.w700,
+                  color: Theme.of(context).colorScheme.onPrimaryFixedVariant),
```
Comment (justinmc) [inline animations/step_01/lib/home_screen.dart:22]: Should we run this whole PR through dartfmt just to make sure that the style is consistent? Up to you.

File: animations/step_03/lib/scoreboard.dart [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return Padding(
+      padding: const EdgeInsets.all(8.0),
+      child: Row(
+        mainAxisAlignment: MainAxisAlignment.center,
+        children: [
+          for (var i = 0; i < totalQuestions; i++)
+            AnimatedStar(
```
Comment (justinmc) [inline animations/step_03/lib/scoreboard.dart:21]: Super nit: Make this private so it's immediately clear to readers that it's located in this file? Not sure if this matters in the context of a codelab though.

File: animations/step_04/lib/flip_effect.dart [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return AnimatedBuilder(
+      animation: _animationWithDelay,
+      builder: (context, child) {
+        return Transform(
+          alignment: Alignment.center,
+          transform: Matrix4.identity()
+            ..rotateX(_animationWithDelay.value * math.pi),
```
Comment (justinmc) [inline animations/step_04/lib/flip_effect.dart:74]: Transformation matrices are a deep topic that we could get into if we have time.

File: animations/step_04/lib/flip_effect.dart [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+          child: _animationController.isAnimating
+              ? _animationWithDelay.value < 0.5
+              ? _previousChild
+              : Transform.flip(flipY: true, child: widget.child)
+              : widget.child,
```
Comment (justinmc) [inline animations/step_04/lib/flip_effect.dart:79]: Would it help the readability if you made this a switch?

`switch ((_animationController.isAnimating, _previousChild)) {...`

**Review by justinmc — APPROVED** [review]
Overall I like it, it builds well upon itself and is pretty interesting as an example. Each concepts seems simple and self-contained enough to be easily undestandable. It was pretty obvious to me what had changed between each step without reading the code first.

Your ValueNotifier "state management" seems simple and good enough to me. Do our other codelabs typically use this approach too?

On all steps, when I tap an answer, the correct answer immediately turns red. I feel like it's not clear to me what's happening, or that the red one is the correct answer. This isn't super important but if we have room to squeeze in more code complexity and/or animations, that might be a good place to do it. 

On my big monitor, some things look tiny, like the stars on the question page. We might want to make them bigger in order to draw attention to them when presenting. Not sure how it will look on the projector though.

For step5, see my PR with OpenContainer animation changes: https://github.com/johnpryan/codelabs/pull/1

File: animations/step_01/lib/home_screen.dart (hunk 1 above)
Comment (domesticmouse) [reply animations/step_01/lib/home_screen.dart:22]: Dart format should be run on this by CI. Which, I suspect, means this needs to be added to the CI scripts...

**Review by domesticmouse — COMMENTED** [review]
LGTM once CI passes

**Comment by johnpryan** [comment]
> Your ValueNotifier "state management" seems simple and good enough to me. Do our other codelabs typically use this approach too?

Yeah this is what we're moving towards recommending in the [state management](https://docs.flutter.dev/get-started/fwe/state-management#using-mvvm-for-your-applications-architecture). You can find some broader architecture recommendations and discussion in [this PR](https://github.com/flutter/samples/pull/2342) too.

> On all steps, when I tap an answer, the correct answer immediately turns red. I feel like it's not clear to me what's happening, or that the red one is the correct answer. This isn't super important but if we have room to squeeze in more code complexity and/or animations, that might be a good place to do it.

Yeah I got this feedback when I showed the app to a friend. It's meant to show the correct answer. Maybe we should animate it!

> On my big monitor, some things look tiny, like the stars on the question page. We might want to make them bigger in order to draw attention to them when presenting. Not sure how it will look on the projector though.

I've been testing on my Android phone. Since we're presenting using Android + scrcpy I think we can keep things as-is but let me know if you run into any major bugs / issues.

> For step5, see my PR with OpenContainer animation changes: johnpryan#1

Merged!

**Comment by johnpryan** [comment]
CI is passing on stable and beta, so I'm going to merge this as-is and follow up with additional changes.

**Pull request merged**